libs/widget-lib/src/cowSwapWidget.ts

```diff
@@ -49,19 +54,36 @@ export function cowSwapWidget(
 
   applyDynamicHeight(iframe, params.height)
 
-  if (provider) {
-    const jsonRpcManager = new JsonRpcManager(contentWindow)
-
-    jsonRpcManager.onConnect(provider)
-  }
+  let jsonRpcManager = updateProvider(contentWindow, null, provider)
 
   iframe.addEventListener('load', () => {
-    updateWidget(params, contentWindow)
+    updateWidget(contentWindow, params)
   })
+  const iFrameCowEventEmitter = new IframeCowEventEmitter(listeners)
+
+  return {
+    updateWidget: (newParams: CowSwapWidgetParams) => updateWidget(contentWindow, newParams),
+    updateListeners: (newListeners?: CowEventListeners) => iFrameCowEventEmitter.updateListeners(newListeners),
+    updateProvider: (newProvider) => {
+      jsonRpcManager = updateProvider(contentWindow, jsonRpcManager, newProvider)
```

Comment:
Is this ok?

This callback updates the jsonRpcManager variable which is inside the fn closure?

Reply:
Why wouldn't it?

I just made an example for you:

```js
callBarman = (() => {
    let count = 0
    return () => {
        count++
        console.log(`Give me ${count} beer${count>1?'s':''}`)
    }
})()
```

<img width="506" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/9b72627d-69e9-4831-91d3-c3694f26d9a4">

Reply:
Alright, I guess.
Just looked weird. Not used to see this pattern.